### PR TITLE
SSE2 & AVX2 std::find & std::count

### LINKED ISSUE #2379: `<xutility>`: optimize `_Find_unchecked` for data sizes other than 1
`_Find_unchecked` is the implementation of `std::find`.

Currently it is optimized for 8 bit size array elements by using `memchr`.

It is possible to expand this optimization for 16 bit size array elements by using `wmemchr`.

For greater size operands, like 32 or 64 bit, it is possible to implement the optimization either, but it takes some manual implementation.

Assuming 32 bit element, the comparison using SSE2 can be made with `_mm_cmpeq_epi32`. Then with `_mm_movemask_epi8`, mask can be extracted. If it is nonzero, `countr_zero` may be used to determine the position of the first match. This only requires SSE2, which is x64 baseline.

With AVX2, there are 256-bit variables available and 64 bit data sizes are possible.

The implementations should probably go to `vector_algorithm.cpp`

### LINKED ISSUE #2384: `<xutility>`: vectorize `std::count`
Relates to #2379

For contiguous ranges, simple types (1,2,4,8 byte integers, maybe also 4,8 bytes float in fast mode) the following vector algorithm is possible (assuming SSE2 and 8-bit type, but applicable to other sizes/vector sizes):

Spread the value to a vector register (`_mm_set1` intrinsics)
Obtain matched bitmask (`_mm_cmpeq_epi8` intrinsic)
Get mask as bits (`_mm_movemask_epi8`) , add them up (`_popcnt`)
Accumulate this result.
Probably hand-coded `popcount` will be inefficient, in this case can apply starting SSE4.2, for which we assume `popcnt` available.

### LINKED ISSUE #2431: <ranges>: Access violation on find in byte range with `unreachable_sentinel`
**Describe the bug**
`ranges::find` with `unreachable_sentinel` may read beyond the found element.
This is due to `memchr` optimization and `memchr` misbehavior reported as DevCom-1615707

**Command-line test case**
```c++
#include <cassert>
#include <ranges>
#include <utility>

#include <Windows.h>

using namespace std;

void* volatile discard;

int main() {
    SYSTEM_INFO si = {};
    GetSystemInfo(&si);

    const size_t page  = si.dwPageSize;
    const size_t alloc = max<size_t>(page * 2, si.dwAllocationGranularity);

    void* p = VirtualAlloc(nullptr, alloc, MEM_RESERVE, PAGE_NOACCESS);
    assert(p != nullptr);
    void* p2 = VirtualAlloc(p, page, MEM_COMMIT, PAGE_READWRITE);
    assert(p2 != nullptr);

    const auto p3    = reinterpret_cast<char*>(p2) + page - 3;
    p3[0]            = '1';
    p3[1]            = '2';
    p3[2]            = '3';
    const auto bingo = ranges::find(p3, unreachable_sentinel, '3');
    assert(bingo == p3 + 2);

    const auto bingo2 = ranges::find(p3 - 16, unreachable_sentinel, '2');
    assert(bingo2 == p3 + 1);

    const auto bingo3 = ranges::find(p3 - 32, unreachable_sentinel, '1');
    assert(bingo3 == p3);

    const auto bingo4 = ranges::find(p3 - 64, unreachable_sentinel, '3');
    assert(bingo4 == p3 + 2);

    VirtualFree(p, 0, MEM_RELEASE);
}
```

**Expected behavior**
No crash. But still with optimization of this case.

**STL version**
Version 17.1.0 Preview 1.1

**Additional context**
Relates to #2379 - in scope of this improvement can also address this problem

### PR DESCRIPTION
**Pull request by AlexGuteniev**
### 📝 Summary

 * Resolve #2379.
   New vector search, controlled by `_USE_STD_VECTOR_ALGORITHMS`. Used also instead of `memchr` optimization for 1-byte element.
 * Fix #2431.
   When `_USE_STD_VECTOR_ALGORITHMS` is off, keep using `memchr` optimization, but only for finite range.
 * Test for #2431.
    For both enabled and disabled `_USE_STD_VECTOR_ALGORITHMS`
 * Resolve #2384.
    New vector count, very similar to vector find, also controlled by `_USE_STD_VECTOR_ALGORITHMS`

### 🏁 Perf benchmark

<details>
<summary><b>Benchmark</b></summary>

```c++
#include <algorithm>
#include <cstdint>
#include <chrono>
#include <iostream>
#include <ranges>
#include <intrin.h>

enum class Kind {
    FindSized,
    FindUnsized,
    Count,
};

template<typename T>
void benchmark_find(T* a, std::size_t max, size_t start, size_t pos, Kind kind, size_t rep) {
    size_t count_expected;
    std::fill_n(a, max, '0');
    if (pos < max && pos >= start) {
        a[pos] = '1';
        count_expected = 1;
    }
    else {
        if (kind == Kind::FindUnsized) {
            abort();
        }
        count_expected = 0;
    }

    auto t1 = std::chrono::steady_clock::now();

    switch (kind)
    {
    case Kind::FindSized:
        for (std::size_t s = 0; s < rep; s++) {
            _ReadWriteBarrier(); // To avoid the compiler moving `memchr` out of loop

            if (std::ranges::find(a + start, a + max, '1') != a + pos) {
                abort();
            }
        }
        break;
    case Kind::FindUnsized:
        for (std::size_t s = 0; s < rep; s++) {
            _ReadWriteBarrier(); // To avoid the compiler moving `memchr` out of loop

            if (std::ranges::find(a + start, std::unreachable_sentinel, '1') != a + pos) {
                abort();
            }
        }
        break;
    case Kind::Count:
        for (std::size_t s = 0; s < rep; s++) {
            _ReadWriteBarrier(); // To avoid the compiler moving `memchr` out of loop

            if (std::ranges::count(a + start, a + max, '1') != count_expected) {
                abort();
            }
        }
        break;
    }

    auto t2 = std::chrono::steady_clock::now();

    const char* op_str = nullptr;
    switch (kind)
    {
    case Kind::FindSized:
        op_str = "find sized";
        break;
    case Kind::FindUnsized:
        op_str = "find unsized";
        break;
    case Kind::Count:
        op_str = "count";
        break;
    }
    std::cout << std::setw(10) << std::chrono::duration_cast<std::chrono::duration<double>>(t2 - t1).count() << "s  -- "
        << "Op " << op_str << " Size " << sizeof(T) << " byte elements, array size " << max
        << " starting at " << start << " found at " << pos << "; " << rep << " repetitions \n";
}


constexpr std::size_t Nmax = 8192;

alignas(64) std::uint8_t    a8[Nmax];
alignas(64) std::uint16_t   a16[Nmax];
alignas(64) std::uint32_t   a32[Nmax];
alignas(64) std::uint64_t   a64[Nmax];

int main()
{
    std::cout << "Vector alg used: " << _USE_STD_VECTOR_ALGORITHMS << "\n";

    benchmark_find(a8, Nmax, 0, 3456, Kind::FindSized, 10000000);
    benchmark_find(a16, Nmax, 0, 3456, Kind::FindSized, 1000000);
    benchmark_find(a32, Nmax, 0, 3456, Kind::FindSized, 1000000);
    benchmark_find(a64, Nmax, 0, 3456, Kind::FindSized, 1000000);

    benchmark_find(a8, Nmax, 0, 3456, Kind::FindUnsized, 1000000);
    benchmark_find(a16, Nmax, 0, 3456, Kind::FindUnsized, 1000000);
    benchmark_find(a32, Nmax, 0, 3456, Kind::FindUnsized, 1000000);
    benchmark_find(a64, Nmax, 0, 3456, Kind::FindUnsized, 1000000);

    benchmark_find(a8, Nmax, 0, 3456, Kind::Count, 1000000);
    benchmark_find(a16, Nmax, 0, 3456, Kind::Count, 1000000);
    benchmark_find(a32, Nmax, 0, 3456, Kind::Count, 1000000);
    benchmark_find(a64, Nmax, 0, 3456, Kind::Count, 1000000);

    std::cout << "Done\n";

    return 0;
}
```
</details>
<details>
<summary><b>Benchmark run and results</b></summary>

```
**********************************************************************
** Visual Studio 2022 Developer Command Prompt v17.1.0-pre.1.1
** Copyright (c) 2021 Microsoft Corporation
**********************************************************************
[vcvarsall.bat] Environment initialized for: 'x64'

C:\Program Files\Microsoft Visual Studio\2022\Preview>cd/d C:\Project\vector_find_benchmark

C:\Project\vector_find_benchmark>set INCLUDE=C:\Project\STL\out\build\x64\out\inc;%INCLUDE%

C:\Project\vector_find_benchmark>set LIB=C:\Project\STL\out\build\x64\out\lib\amd64;%LIB%

C:\Project\vector_find_benchmark>set PATH=C:\Project\STL\out\build\x64\out\bin\amd64;%PATH%

C:\Project\vector_find_benchmark>cl /O2 /std:c++latest /EHsc /D_USE_STD_VECTOR_ALGORITHMS=0 /nologo vector_find_benchmark.cpp
vector_find_benchmark.cpp

C:\Project\vector_find_benchmark>vector_find_benchmark.exe
Vector alg used: 0
  0.123933s  -- Op find sized Size 1 byte elements, array size 8192 starting at 0 found at 3456; 1000000 repetitions
  0.911876s  -- Op find sized Size 2 byte elements, array size 8192 starting at 0 found at 3456; 1000000 repetitions
  0.917381s  -- Op find sized Size 4 byte elements, array size 8192 starting at 0 found at 3456; 1000000 repetitions
  0.975016s  -- Op find sized Size 8 byte elements, array size 8192 starting at 0 found at 3456; 1000000 repetitions
    1.1216s  -- Op find unsized Size 1 byte elements, array size 8192 starting at 0 found at 3456; 1000000 repetitions
   1.14476s  -- Op find unsized Size 2 byte elements, array size 8192 starting at 0 found at 3456; 1000000 repetitions
    1.1659s  -- Op find unsized Size 4 byte elements, array size 8192 starting at 0 found at 3456; 1000000 repetitions
   1.11486s  -- Op find unsized Size 8 byte elements, array size 8192 starting at 0 found at 3456; 1000000 repetitions
    4.2511s  -- Op count Size 1 byte elements, array size 8192 starting at 0 found at 3456; 1000000 repetitions
   4.23966s  -- Op count Size 2 byte elements, array size 8192 starting at 0 found at 3456; 1000000 repetitions
   4.26283s  -- Op count Size 4 byte elements, array size 8192 starting at 0 found at 3456; 1000000 repetitions
   3.19778s  -- Op count Size 8 byte elements, array size 8192 starting at 0 found at 3456; 1000000 repetitions
Done

C:\Project\vector_find_benchmark>cl /O2 /std:c++latest /EHsc /D_USE_STD_VECTOR_ALGORITHMS=1 /nologo vector_find_benchmark.cpp
vector_find_benchmark.cpp

C:\Project\vector_find_benchmark>vector_find_benchmark.exe
Vector alg used: 1
 0.0566276s  -- Op find sized Size 1 byte elements, array size 8192 starting at 0 found at 3456; 1000000 repetitions
 0.0979886s  -- Op find sized Size 2 byte elements, array size 8192 starting at 0 found at 3456; 1000000 repetitions
  0.185901s  -- Op find sized Size 4 byte elements, array size 8192 starting at 0 found at 3456; 1000000 repetitions
  0.357151s  -- Op find sized Size 8 byte elements, array size 8192 starting at 0 found at 3456; 1000000 repetitions
  0.040667s  -- Op find unsized Size 1 byte elements, array size 8192 starting at 0 found at 3456; 1000000 repetitions
 0.0683121s  -- Op find unsized Size 2 byte elements, array size 8192 starting at 0 found at 3456; 1000000 repetitions
  0.239233s  -- Op find unsized Size 4 byte elements, array size 8192 starting at 0 found at 3456; 1000000 repetitions
  0.247233s  -- Op find unsized Size 8 byte elements, array size 8192 starting at 0 found at 3456; 1000000 repetitions
  0.185978s  -- Op count Size 1 byte elements, array size 8192 starting at 0 found at 3456; 1000000 repetitions
  0.349529s  -- Op count Size 2 byte elements, array size 8192 starting at 0 found at 3456; 1000000 repetitions
  0.722243s  -- Op count Size 4 byte elements, array size 8192 starting at 0 found at 3456; 1000000 repetitions
   1.59221s  -- Op count Size 8 byte elements, array size 8192 starting at 0 found at 3456; 1000000 repetitions
Done
```
</details>
<details>
<summary><b>Results table</b></summary>

  size              |  before   |  after     |
--------------------|-----------| -----------|
find sized 1 byte   | 0.123933s | 0.0566276s |
find sized 2 bytes  | 0.911876s | 0.0979886s |
find sized 4 bytes  | 0.917381s |  0.185901s |
find sized 8 bytes  | 0.975016s |  0.357151s |
find unsized 1 byte |   1.1216s |  0.040667s |
find unsized 2 bytes|  1.14476s | 0.0683121s |
find unsized 4 bytes|   1.1659s |  0.239233s |
find unsized 8 bytes|  1.11486s |  0.247233s |
count 1 byte        |   4.2511s |  0.185978s |
count 2 bytes       |  4.23966s |  0.349529s |
count 4 bytes       |  4.26283s |  0.722243s |
count 8 bytes       |  3.19778s |   1.59221s |
</details>

**TL;DR:** all cases have significant benefit. **Decimal order of magnitude for some cases**.

When AVX is disabled, the benefit becomes small for the following cases:
 * 64-bit find sized Size 1 - as `memchr` is also SSE2, and my version only slightly beats it, as they are very similar
 * 64-bit Size 8 all algorithms - as vector only contains two elements, overhead for vectorization becomes comparable to gain

Still no regression found on tested cases.

### ⚖️ Size impact

The change adds more code, it also overrides (disables) `/Os` option for `vector_algorithm.cpp` via pragma.
DLLs and PDBs for them are not affected. Static libraries are affected. 
**The impact is negligible for static libs, small for import libs**.

<details>
<summary><b>Table</b></summary>

|File name      |Size before  |Size after  |
|:---           |         ---:|       ---: | 
|libcpmt.lib    |  31,713,452 | 31,801,394 |
|libcpmt1.lib   |  32,551,896 | 32,641,100 |
|libcpmtd.lib   |  33,254,860 | 33,352,526 |
|libcpmtd0.lib  |  32,095,116 | 32,192,820 |
|libcpmtd1.lib  |  32,987,834 | 33,085,806 |
|msvcprt.lib    |   1,038,290 |  1,119,228 |
|msvcprtd.lib   |   1,045,802 |  1,133,746 |
|stl_asan.lib   |       3,030 |      3,030 |
</details>

### ✔️ Test coverage
 * Expand **Dev11_0316853_find_memchr_optimization** to test new types, mostly **limits** test
 * Add **GH_002431_byte_range_find_with_unreachable_sentinel** to test for as if sequential reading when the size is not set
 * Expand **VSO_0000000_vector_algorithms** to test newly vectorized cases with various sizes

###  ⚠️  ARM64EC Note
I didn't try to make it working on ARM64EC. I'm not sure how to make it correctly. 
 * From other parallel algorithm looks like I should enable SSE, but not AVX - see `_M_ARM64EC` usage 
 * From `<bit>` it looks like that ARM version of `popcount` is used instead of x86.
 * Finally, parallel algorithms are disabled under `_M_HYBRID`, which might have to do something with ARM64EC
It would be good to have #2310 addressed, so that these changes can be tested.

### CODE REVIEW

File: stl/inc/xutility [hunk 1]
```diff
@@ -5087,75 +5108,149 @@ _NODISCARD constexpr auto lexicographical_compare_three_way(
...
+        using _Elem = _Iter_value_t<_InIt>;
+        _STL_INTERNAL_STATIC_ASSERT(
+            sizeof(_Elem) == 1 || sizeof(_Elem) == 2 || sizeof(_Elem) == 4 || sizeof(_Elem) == 8);
+        if constexpr (is_same_v<_Elem, bool>) {
+            return _Val == true || _Val == false;
+        } else if constexpr (is_signed_v<_Elem>) {
+            if constexpr (is_signed_v<_Ty>) {
+                // signed _Elem, signed _Ty
+                if constexpr (sizeof(_Elem) == 1) {
+                    return SCHAR_MIN <= _Val && _Val <= SCHAR_MAX;
```
Comment (AlexGuteniev) [inline stl/inc/xutility:5132]: To avoid braching on size here, compute limits using two's complement, shift and `CHAR_BIT`

**Review by fsb4000 — APPROVED** [review]
It looks good to me but I'm not an expert :)

**Comment by cbezault** [comment]
Honestly I don't think we should even use the SSE instructions under ARM64EC.

File: tests/std/tests/GH_002431_byte_range_find_with_unreachable_sentinel/test.cpp [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    SYSTEM_INFO si = {};
+    GetSystemInfo(&si);
+
+    const size_t page  = si.dwPageSize;
+    const size_t alloc = max<size_t>(page * 2, si.dwAllocationGranularity);
+
+    void* p = VirtualAlloc(nullptr, alloc, MEM_RESERVE, PAGE_NOACCESS);
+    assert(p != nullptr);
+    void* p2 = VirtualAlloc(p, page, MEM_COMMIT, PAGE_READWRITE);
+    assert(p2 != nullptr);
```
Comment (barcharcraz) [inline tests/std/tests/GH_002431_byte_range_find_with_unreachable_sentinel/test.cpp:55]: Why bother doing this in two steps instead of reserving and committing in one?

**Review by barcharcraz — CHANGES_REQUESTED** [review]

File: stl/src/vector_algorithms.cpp [hunk 3]
```diff
@@ -451,6 +452,358 @@ __declspec(noalias) void __cdecl __std_reverse_copy_trivially_copyable_8(
...
+        constexpr size_t _Vector_pad_mask = 0xF;
+        constexpr unsigned _Full_mask     = 0xFFFF;
+
+        const __m128i _Comparand  = _Traits::_Set_sse(_Val);
+        const intptr_t _Pad_start = reinterpret_cast<intptr_t>(_First) & _Vector_pad_mask;
+        const unsigned _Mask      = (_Full_mask << _Pad_start);
+        _Advance_bytes(_First, -_Pad_start);
+
+        __m128i _Data   = _mm_load_si128(static_cast<const __m128i*>(_First));
+        unsigned _Bingo = static_cast<unsigned>(_mm_movemask_epi8(_Traits::_Cmp_sse(_Data, _Comparand)));
```
[the remark this answers is not in the document]
Comment (AlexGuteniev) [reply stl/src/vector_algorithms.cpp:639]: 16 actually used bits, high 16 are zero.
Avoiding 16-bit type for more effecient codegen (avoid widening where it is not needed.

File: stl/src/vector_algorithms.cpp [hunk 4]
```diff
@@ -451,6 +452,358 @@ __declspec(noalias) void __cdecl __std_reverse_copy_trivially_copyable_8(
...
+
+        if ((_Bingo &= _Mask) != 0) {
+            unsigned long _Offset;
+            _BitScanForward(&_Offset, _Bingo);
```
[the remark this answers is not in the document]
Comment (AlexGuteniev) [reply stl/src/vector_algorithms.cpp:643]: Remember that using `tzcnt` without branch for old cpu caused issues and was reverted. Don't want to use a `tzcnt`/ `bsf` branch here, think that smaller code is bettee here.

File: stl/src/vector_algorithms.cpp [hunk 5]
```diff
@@ -451,6 +452,358 @@ __declspec(noalias) void __cdecl __std_reverse_copy_trivially_copyable_8(
...
+                return _First;
+            }
+
+            _Advance_bytes(_First, 32);
+        } while (_First != _Stop_at);
+        Size_bytes &= 0x1F;
+    }
+
+    const size_t _Sse_size = Size_bytes & ~size_t{0xF};
+    if (_Sse_size != 0 && _Traits::_Sse_available()) {
```
[the remark this answers is not in the document]
Comment (AlexGuteniev) [reply stl/src/vector_algorithms.cpp:692]: No, it is on par (no clear advantage or disadvantage).
The main reason I dis this way is symmetry with preexisting algs.

File: stl/src/vector_algorithms.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (barcharcraz) [reply stl/src/vector_algorithms.cpp:639]: Allright, that seems correct, and that's what I meant, the extra bits are zeroed.

File: stl/src/vector_algorithms.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (barcharcraz) [reply stl/src/vector_algorithms.cpp:643]: well, we already have the branch for zero, but I agree, more trouble than it's worth.

File: stl/src/vector_algorithms.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (barcharcraz) [reply stl/src/vector_algorithms.cpp:692]: Yeah, I guess it might save a few instruction cache lines to skip it, which might not show up in benchmarks, but I'm OK with continuing to do both.

File: stl/CMakeLists.txt [hunk 6]
```diff
@@ -265,6 +265,7 @@ set(IMPLIB_SOURCES
 set(DLL_SOURCES
     ${CMAKE_CURRENT_LIST_DIR}/src/dllmain.cpp
     ${CMAKE_CURRENT_LIST_DIR}/src/instances.cpp
+    ${CMAKE_CURRENT_LIST_DIR}/src/vector_algorithms.cpp
```
[the remark this answers is not in the document]
Comment (AlexGuteniev) [reply stl/CMakeLists.txt:268]: Labelling `affects redist` is appropriate.
I think export validation wouldn't fail, as these are not exported.

File: stl/inc/xutility [hunk 7]
```diff
@@ -5215,24 +5310,47 @@ namespace ranges {
...
+                    return _RANGES next(_STD move(_First), _Result - _First_ptr);
                 }
-                const auto _First_ptr = _STD to_address(_First);
-                const auto _Result    = static_cast<remove_reference_t<_Iter_ref_t<_It>>*>(
-                    _CSTD memchr(_First_ptr, static_cast<unsigned char>(_Val), _Count));
-                if (_Result) {
-                    if constexpr (is_pointer_v<_It>) {
-                        return _Result;
+#else // ^^^ _USE_STD_VECTOR_ALGORITHMS ^^^ / vvv not _USE_STD_VECTOR_ALGORITHMS vvv
+                if constexpr (sizeof(_Iter_value_t<_It>) == 1 && _Is_sized /* TRANSITION, DevCom-1615707 */) {
```
[the remark this answers is not in the document]
Comment (AlexGuteniev) [reply stl/inc/xutility:5331]: Possibly, but I can't check it. Does this issue have public reference?

File: stl/src/vector_algorithms.cpp [hunk 8]
```diff
@@ -451,6 +452,358 @@ __declspec(noalias) void __cdecl __std_reverse_copy_trivially_copyable_8(
...
+const void* __stdcall __std_find_trivial(const void* _First, const void* _Last, _Ty _Val) noexcept {
+    size_t Size_bytes = _Byte_length(_First, _Last);
+
+    const size_t _Avx_size = Size_bytes & ~size_t{0x1F};
+    if (_Avx_size != 0 && _Use_avx2()) {
+        const __m256i _Comparand = _Traits::_Set_avx(_Val);
+        const void* _Stop_at     = _First;
+        _Advance_bytes(_Stop_at, _Avx_size);
+        do {
+            const __m256i _Data = _mm256_loadu_si256(static_cast<const __m256i*>(_First));
```
[the remark this answers is not in the document]
Comment (AlexGuteniev) [reply stl/src/vector_algorithms.cpp:677]: I'm afraid I don't understand the question.

For all algorithms here, the buffer isn't assumed to be 32 or 16 bytes aligned, so aligned instruction may fail.

The only exception is unsized algorithms. They align the buffer before processing to avoid hard page fault, so they can rely on alignment.

Can use unalignned instruction for unsized algorithms too. But here we must use unaligned.

File: stl/src/vector_algorithms.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (AlexGuteniev) [reply stl/src/vector_algorithms.cpp:692]: I understand the instruction cache concern, but I failed to justify it by failing to provide a benchmark where the effect is above the noise ratio.

File: tests/std/tests/GH_002431_byte_range_find_with_unreachable_sentinel/test.cpp (hunk 2 above)
Comment (AlexGuteniev) [reply tests/std/tests/GH_002431_byte_range_find_with_unreachable_sentinel/test.cpp:55]: To create page guard where we reserve at least two pages, but commit one page and make access violation on accessing beyond the range mandatory

File: stl/inc/xutility [hunk 9]
```diff
@@ -5100,43 +5177,48 @@ _NODISCARD constexpr bool _Within_limits(const _InIt&, const _Ty& _Val) {
...
 }
 
 template <class _Iter, class _Ty>
-_INLINE_VAR constexpr bool _Memchr_in_find_is_safe =
-    _Iterator_is_contiguous<_Iter>&&
-        disjunction_v<conjunction<is_integral<_Ty>, _Is_character_or_bool<_Iter_value_t<_Iter>>>
+_INLINE_VAR constexpr bool _Vector_alg_in_find_is_safe =
+    _Iterator_is_contiguous<_Iter>&& disjunction_v<conjunction<is_integral<_Ty>, is_integral<_Iter_value_t<_Iter>>>,
+        conjunction<disjunction<is_pointer<_Ty>, is_null_pointer<_Ty>>,
+            disjunction<is_pointer<_Iter_value_t<_Iter>>, is_null_pointer<_Iter_value_t<_Iter>>>>
```
[the remark this answers is not in the document]
Comment (AlexGuteniev) [reply stl/inc/xutility:5237]: I don't understand the problem

File: stl/inc/xutility [hunk 10]
```diff
@@ -5100,43 +5177,48 @@ _NODISCARD constexpr bool _Within_limits(const _InIt&, const _Ty& _Val) {
...
         }
     }
 }
 
 template <class _Iter, class _Ty>
-_INLINE_VAR constexpr bool _Memchr_in_find_is_safe =
-    _Iterator_is_contiguous<_Iter>&&
-        disjunction_v<conjunction<is_integral<_Ty>, _Is_character_or_bool<_Iter_value_t<_Iter>>>
+_INLINE_VAR constexpr bool _Vector_alg_in_find_is_safe =
+    _Iterator_is_contiguous<_Iter>&& disjunction_v<conjunction<is_integral<_Ty>, is_integral<_Iter_value_t<_Iter>>>,
```
[the remark this answers is not in the document]
Comment (AlexGuteniev) [reply stl/inc/xutility:5235]: What comment?

File: tests/std/tests/Dev11_0316853_find_memchr_optimization/test.cpp [hunk 11]
```diff
@@ -54,23 +154,34 @@ int main() {
...
 
     { // Optimization inapplicable due to bogus element type and bogus value type
         vector<string> v;
         v.push_back("cute");
+        v.push_back("cute");
         v.push_back("fluffy");
         v.push_back("kittens");
 
-        assert(find(v.begin(), v.end(), "fluffy") == v.begin() + 1);
+        assert(find(v.begin(), v.end(), "fluffy") == v.begin() + 2);
```
[the remark this answers is not in the document]
Comment (AlexGuteniev) [reply tests/std/tests/Dev11_0316853_find_memchr_optimization/test.cpp:174]: How? Would we directly test `_Vector_alg_in_find_is_safe`? Or intercept calls like `__std_count_trivial` via preprocessor, and see if they are called or not?

File: stl/inc/xutility (hunk 10 above)
[the remark this answers is not in the document]
Comment (barcharcraz) [reply stl/inc/xutility:5235]: well, when reviewing I got out notepad and wrote this down with slightly abbreviated type trait names and normal logical operators (||,&&,())

It's not a huge deal.

File: stl/src/vector_algorithms.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (barcharcraz) [reply stl/src/vector_algorithms.cpp:687]: yeah, I mean that's always the trouble with i$ concerns, targeted benchmarks almost never show the issue.

Anyway, it's fine to keep this as-is.

File: tests/std/tests/Dev11_0316853_find_memchr_optimization/test.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (barcharcraz) [reply tests/std/tests/Dev11_0316853_find_memchr_optimization/test.cpp:174]: I think just calling `_Vector_alg_find_is_safe` is probably the simplest, intercepting calls sounds like more pain than it's worth.

File: tests/std/tests/GH_002431_byte_range_find_with_unreachable_sentinel/test.cpp (hunk 2 above)
Comment (barcharcraz) [reply tests/std/tests/GH_002431_byte_range_find_with_unreachable_sentinel/test.cpp:55]: oh I see.

File: stl/src/vector_algorithms.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (AlexGuteniev) [reply stl/src/vector_algorithms.cpp:672]: Ah, I think I got it. Note that the usual loop is compiled anyway. `if constexpr (_Vector_alg_in_find_is_safe<_InIt, _Ty>)` don't even have `else`. So the usual loop checks types.

File: stl/inc/xutility (hunk 9 above)
[the remark this answers is not in the document]
Comment (AlexGuteniev) [reply stl/inc/xutility]: However there might be a problem for base/derived pointers.... Need to address that and add a test for it!

[the remark this answers is not in the document]
Comment (AlexGuteniev) [reply stl/inc/xutility]: Ok, will use for now only for *exactly the same* pointers. 
Full pointer optimization (with proper pointer casting, including virtual inheritance) can be implemented in a next PR.

File: stl/CMakeLists.txt (hunk 6 above)
[the remark this answers is not in the document]
Comment (StephanTLavavej) [reply stl/CMakeLists.txt:270]: We'd like to understand why this needs to be added to the DLL (for internal usage) in the first place - are there compiler errors if we don't do this? (@CaseyCarter noted that if we're calling `find/count` in the DLL, then that could be the issue - I didn't expect that would be the case but I tend to forget everything we do in the DLL.)

File: tests/std/tests/Dev11_0316853_find_memchr_optimization/test.cpp [hunk 12]
```diff
@@ -29,22 +32,125 @@ bool operator==(int x, const Cat& c) {
...
+
         vector<signed char> v;
         v.push_back(22);
         v.push_back(33);
         v.push_back(-1);
         v.push_back(44);
+        v.push_back(44);
         v.push_back(55);
 
+#ifdef __cpp_lib_concepts
```
Comment (barcharcraz) [inline tests/std/tests/Dev11_0316853_find_memchr_optimization/test.cpp:139]: why are these guarded by concepts?

**Review by barcharcraz — APPROVED** [review]
Things are looking good, I still want to run the tests for this on ARM64 to make sure we're not breaking things there.

File: stl/CMakeLists.txt (hunk 6 above)
[the remark this answers is not in the document]
Comment (StephanTLavavej) [reply stl/CMakeLists.txt:270]: Ah, this is the reason: https://github.com/microsoft/STL/blob/27877181dc50fc5f0dc9d679703437eb105e2b9f/stl/inc/xlocnum#L78-L83

File: tests/std/tests/Dev11_0316853_find_memchr_optimization/test.cpp (hunk 12 above)
Comment (AlexGuteniev) [reply tests/std/tests/Dev11_0316853_find_memchr_optimization/test.cpp:139]: `std::vector` iterators are not detected as contiguous without concepts.
We might look into it, and expand optimization for pre-C++20, but even if we do, i prefer it i a next PR.

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]

**Comment by StephanTLavavej** [comment]
This is amazing, thanks! :heart_eyes_cat: I think this will be ready to merge after resolving the issues I found - please let me know if you'd like me to push changes for them (none were in the core vectorized machinery - the number was just large enough that I didn't feel comfortable pushing changes without asking first).

**Comment by AlexGuteniev** [comment]
Yes please go ahead!

File: stl/CMakeLists.txt [hunk 13]
```diff
@@ -267,6 +267,7 @@ set(IMPLIB_SOURCES
...
```
[the remark this answers is not in the document]
Comment (StephanTLavavej) [reply stl/CMakeLists.txt:270]: I'm going to disable vector algorithms when building the DLL. (Rationale: Not a regression, there's just one `find()` call and it's not perf-sensitive to my knowledge.)

File: stl/inc/xutility [hunk 14]
```diff
@@ -5116,65 +5197,86 @@ _NODISCARD constexpr bool _Within_limits(const _InIt&, const _Ty& _Val) {
...
             }
         }
     }
 }
 
-template <class _Iter, class _Ty>
-_INLINE_VAR constexpr bool _Memchr_in_find_is_safe =
+// _Iterator_is_contiguous<_Iter> && (is_integral<_Ty> && is_integral<_Val> ||
+// is_pointer<_Ty> && is_pointer<_Val> || is_null_pointer<_Val>) ||
+// is_same<_Ty, byte> && is_same<_Ty, _Val>
```
[the remark this answers is not in the document]
Comment (StephanTLavavej) [reply stl/inc/xutility:5235]: I rearranged and narrated this. It should be a lot easier to understand now.

File: stl/src/vector_algorithms.cpp [hunk 15]
```diff
@@ -450,6 +451,354 @@ __declspec(noalias) void __cdecl __std_reverse_copy_trivially_copyable_8(
...
+static const void* _Find_trivial_tail(const void* _First, const void* _Last, _Ty _Val) {
+    auto _Ptr = static_cast<const _Ty*>(_First);
+    while (_Ptr != _Last && *_Ptr != _Val) {
+        ++_Ptr;
+    }
+    return _Ptr;
+}
+
+template <class _Ty>
+__declspec(noalias) size_t _Count_trivial_tail(const void* _First, const void* _Last, size_t _Current, _Ty _Val) {
```
Comment (StephanTLavavej) [inline stl/src/vector_algorithms.cpp:475]: (Going to push changes for this, but wanted to record why - I had a comment for this earlier but somehow lost it.)

Some but not all of the functions outside of `extern "C"` are `static`. Now that this is gaining types like `struct _Find_traits_1` (which can't be `static`), I think we should change to using an unnamed namespace for all support machinery here. As this is injected into the import lib, we should be maximally clean regarding linkage.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
To fix an internal build break (where the new vector algorithms wouldn't compile for ARM64EC), I've completely disabled the vector algorithms for ARM64EC. This includes the existing attempt to enable the SSE2 parts of `reverse` and `reverse_copy`. When investigating how to add ARM64EC build coverage, @cbezault found he had to do the same thing.

We can explore re-enabling the vector algorithms (and extending them to `find`/`count`) for ARM64EC in the future, but for now I believe that this is the lowest-risk option.

**Comment by StephanTLavavej** [comment]
Thanks for this major performance improvement and bugfix! :rocket: :lady_beetle: :heart_eyes_cat:

**Comment by cbezault** [comment]
Disabling these algorithms for ARM64EC will be a major perf degradation for x64 binaries running on ARM64 devices 😢
